### PR TITLE
Changing RegistrySlice to have ids better determined

### PR DESCRIPTION
**Pull request by rob-c**
Small cleanup of the use of keys within ```RegistrySlice``` This was 'fixed' as part of making sure ```SubjobXMLList``` worked with subjobs  properly from a user perspective but I've realised it should be used everywhere in the class as it's more correct.

I've checked ```SubjobXMLList``` and it now correctly supports the keys method in a way which doesn't cause problems.

I've tried to make sure that the ```keys()``` method isn't called in lists etc as this is a locked list which if it's called many times behind the scenes due to some poor code there will at least be a performance cost due to the number of calls to the function with the constant (un)locking.

### CODE REVIEW

**Comment by rob-c** [comment]
retest this please

**Comment by rob-c** [comment]
retest this please

**Comment by rob-c** [comment]
Any objections to merging this, it's mainly a cleanup now that ```SubJobXMLList``` now supports keys correctly.

File: python/Ganga/GPIDev/Lib/Registry/RegistrySlice.py [hunk 1]
```diff
@@ -304,7 +301,8 @@ def copy(self, keep_going):
         return this_slice
 
     def __contains__(self, j):
-        return j.id in self.objects.keys()
+        allObjK = self.objects.keys()
+        return j.id in allObjK
```
Comment (milliams) [inline python/Ganga/GPIDev/Lib/Registry/RegistrySlice.py]: Why was this split onto two lines? Likewise with the two similar changes below. I feel it makes the code slightly more verbose for no gain.

Comment (rob-c) [inline python/Ganga/GPIDev/Lib/Registry/RegistrySlice.py]: This was to reduce the objects.keys potentially being called multiple times. I'm no expert on when python evaluates commands on inline stuff like this and I was going for clarity that this was called just once.

Comment (milliams) [inline python/Ganga/GPIDev/Lib/Registry/RegistrySlice.py]: The code is exactly the same before and after the change. When doing:
```python
j.id in self.objects.keys()
```
it only calls `self.objects.keys()` once.

Comment (rob-c) [inline python/Ganga/GPIDev/Lib/Registry/RegistrySlice.py]: Is this always the case in for loops in addition to the ```in``` test?

Comment (milliams) [inline python/Ganga/GPIDev/Lib/Registry/RegistrySlice.py]: Yes, the thing `Y` in a `for X in Y:` will be assigned to a temporary internal variable and have `iter()` called on it which then has `next()` called on it until it raises a `StopIteration` exception. In this case, `keys()` returns a `list` which which, when passed to `iter()` returns a `listiterator` object. Something like:
```python
for X in Y:
    ... something using X ...

→

_Y = iter(Y)
while True:
    try:
        X = next(_Y)
        ... something using X ...
    except StopIteration:
        break
```
so `Y` is only accessed once.

Comment (rob-c) [inline python/Ganga/GPIDev/Lib/Registry/RegistrySlice.py]: Yes but we can't do ```for x in y``` where y is a ```SubJobXMLList``` in which case ```.keys()``` allows for lazy loading

Comment (milliams) [inline python/Ganga/GPIDev/Lib/Registry/RegistrySlice.py]: I'm afraid I don't understand.

Comment (rob-c) [inline python/Ganga/GPIDev/Lib/Registry/RegistrySlice.py]: This was just a comment as to why we can't drop the use of some ```.keys``` calls, but I hope this is clear in the revision after this which touches less code.

**Comment by rob-c** [comment]
Any objections to squashing and merging this if the tests pass?

**Comment by milliams** [comment]
:+1:

**Pull request merged**